File: backend/app/analysis/reps.py
"""Repetition detection.

Finds squat repetitions in the hip-height signal produced by `angles.py`.

**The approach: a hysteresis state machine over adaptive thresholds.**

The naive version — count every time the hip drops below some fixed height — has
two fatal problems. First, a fixed height is meaningless: it depends on the
lifter, the camera, and the frame. Second, a single threshold flickers. A hip
hovering right at the boundary crosses it repeatedly and registers a dozen reps
where there was one.

Both are solved here. Thresholds are derived from the range of hip travel *in
this particular video*, so they adapt to the lifter automatically. And there are
two of them — a lower one to enter the descent, a higher one to close the rep —
so the signal must genuinely traverse the gap between them to advance. That gap
is the hysteresis band, and it makes flickering impossible.
"""

from __future__ import annotations

from enum import Enum, auto

from app.analysis.geometry import linear_scale
from app.analysis.smoothing import percentile
from app.analysis.types import AngleSeries, Rep, ViewOrientation
from app.config import Settings
from app.logging_config import get_logger

logger = get_logger(__name__)


class _State(Enum):
    """Where the lifter is in the movement."""

    STANDING = auto()
    DESCENDING = auto()
    ASCENDING = auto()


def detect_reps(angles: AngleSeries, settings: Settings) -> list[Rep]:
    """Find every repetition in the clip.

    Returns an empty list — rather than raising — when the video contains no
    detectable movement. "No reps" is a legitimate outcome that the coaching
    layer reports on; it is not an error.
    """
    if len(angles) == 0:
        return []

    signal = angles.hip_height
    baseline = percentile(signal, 90)
    bottom_reference = percentile(signal, 10)

    if baseline is None or bottom_reference is None:
        logger.info("No usable hip-height signal; reporting zero reps")
        return []

    travel = baseline - bottom_reference

    # Percentiles rather than min/max: one mis-tracked frame at an extreme would
    # otherwise define the entire range that every threshold derives from.
    if travel < settings.min_rep_range:
        logger.info(
            "Hip travel %.3f is below the %.3f minimum; reporting zero reps",
            travel,
            settings.min_rep_range,
        )
        return []

    descend_threshold = baseline - settings.rep_descent_fraction * travel
    ascend_threshold = baseline - settings.rep_ascent_fraction * travel

    candidates = _scan(signal, angles, descend_threshold, ascend_threshold)
    reps = _finalise(candidates, angles, settings)

    logger.info(
        "Detected %d repetitions (travel=%.3f, thresholds %.3f/%.3f)",
        len(reps),
        travel,
        descend_threshold,
        ascend_threshold,
    )
    return reps


def _scan(
    signal: list[float | None],
    angles: AngleSeries,
    descend_threshold: float,
    ascend_threshold: float,
) -> list[tuple[int, int, int]]:
    """Walk the signal and emit ``(start, bottom, end)`` frame triples.

    The state machine:

    - **STANDING** -> DESCENDING once the hip falls below ``descend_threshold``.
      The rep's *start* is backdated to when the hip last left the standing
      band, so the recorded descent includes the whole movement rather than
      only the part below the threshold.
    - **DESCENDING** -> ASCENDING once the hip begins rising again, tracked by
      remembering the lowest point seen.
    - **ASCENDING** -> STANDING once the hip rises back above
      ``ascend_threshold``, which closes the rep.

    Untracked frames (None) do not advance the machine. A rep in progress
    survives a brief dropout rather than being split in two.
    """
    triples: list[tuple[int, int, int]] = []

    state = _State.STANDING
    start_frame = 0
    bottom_frame = 0
    bottom_value = float("inf")
    # Most recent frame at which the lifter was unambiguously standing; used to
    # backdate the start of a descent.
    last_standing_frame = 0

    for index, value in enumerate(signal):
        if value is None:
            continue

        if state is _State.STANDING:
            if value >= ascend_threshold:
                last_standing_frame = index
            elif value < descend_threshold:
                state = _State.DESCENDING
                start_frame = last_standing_frame
                bottom_frame = index
                bottom_value = value

        elif state is _State.DESCENDING:
            if value < bottom_value:
                bottom_value = value
                bottom_frame = index
            elif value > bottom_value:
                # The hip has started back up: the bottom is behind us.
                state = _State.ASCENDING

        elif state is _State.ASCENDING:
            if value < bottom_value:
                # Dipped again before locking out — treat it as still the same
                # descent rather than closing a rep the lifter has not finished.
                bottom_value = value
                bottom_frame = index
                state = _State.DESCENDING
            elif value >= ascend_threshold:
                triples.append((start_frame, bottom_frame, index))
                state = _State.STANDING
                last_standing_frame = index
                bottom_value = float("inf")

    # A rep still in progress at the final frame is deliberately discarded: the
    # lifter never returned to lockout, so its duration and ascent are unknown.
    if state is not _State.STANDING:
        logger.debug("Video ended mid-repetition; incomplete rep discarded")

    return triples


def _finalise(
    triples: list[tuple[int, int, int]],
    angles: AngleSeries,
    settings: Settings,
) -> list[Rep]:
    """Turn frame triples into `Rep` objects, discarding implausible ones."""
    reps: list[Rep] = []

    for start, bottom, end in triples:
        start_time = angles.timestamps_s[start]
        end_time = angles.timestamps_s[end]

        # Anything faster than a person can physically squat is tracking
        # jitter that survived smoothing, not a repetition.
        if end_time - start_time < settings.min_rep_duration_s:
            logger.debug(
                "Discarding %.2fs candidate below the %.2fs minimum",
                end_time - start_time,
                settings.min_rep_duration_s,
            )
            continue

        reps.append(_build_rep(len(reps) + 1, start, bottom, end, angles, settings))

    return reps


def _build_rep(
    index: int,
    start: int,
    bottom: int,
    end: int,
    angles: AngleSeries,
    settings: Settings,
) -> Rep:
    """Measure one repetition.

    Depth is taken at the bottom frame. Lean and asymmetry are taken as the
    worst value across the whole rep, because a lifter who is upright at the top
    and folded at the bottom has a lean problem — averaging would hide it.
    """
    left_at_bottom = _window_min(angles.left_knee_deg, start, end)
    right_at_bottom = _window_min(angles.right_knee_deg, start, end)

    knee_values = [v for v in (left_at_bottom, right_at_bottom) if v is not None]
    min_knee = min(knee_values) if knee_values else None

    # Asymmetry is the worst instantaneous gap between the two knees across the
    # rep, not the gap between each leg's independent minimum. Each leg reaches
    # its deepest point at a slightly different frame, so differencing the two
    # minima cancels a genuine one-sided shift: a lifter visibly loading one leg
    # measured 0.2 degrees of "asymmetry" that way, because both knees passed
    # through the same minimum at different moments. Comparing frame by frame and
    # keeping the worst gap matches the same "worst across the rep" intent used
    # for lean above.
    #
    # Asymmetry needs a front-on camera. Side-on, the far leg is occluded and
    # tracked in a small, noisy fraction of frames — comparing it against the
    # near leg produced ~38 degrees of "asymmetry" on footage with none, which
    # is worse than saying nothing.
    asymmetry = (
        _window_max_gap(angles.left_knee_deg, angles.right_knee_deg, start, end)
        if angles.view is ViewOrientation.FRONT
        else None
    )

    depth_percent = None
    if min_knee is not None:
        depth_percent = (
            linear_scale(
                min_knee,
                settings.standing_knee_angle_deg,
                settings.parallel_knee_angle_deg,
            )
            * 100.0
        )

    offset = _window_max(angles.hip_knee_offset, start, end)

    return Rep(
        index=index,
        start_frame=start,
        bottom_frame=bottom,
        end_frame=end,
        start_time_s=angles.timestamps_s[start],
        bottom_time_s=angles.timestamps_s[bottom],
        end_time_s=angles.timestamps_s[end],
        min_knee_angle_deg=min_knee,
        min_left_knee_deg=left_at_bottom,
        min_right_knee_deg=right_at_bottom,
        min_hip_angle_deg=_window_min(angles.hip_deg, start, end),
        max_torso_lean_deg=_window_max(angles.torso_lean_deg, start, end),
        knee_asymmetry_deg=asymmetry,
        depth_percent=depth_percent,
        hip_below_knee=offset is not None and offset >= 0.0,
    )


def _window_min(values: list[float | None], start: int, end: int) -> float | None:
    """Smallest present value in ``[start, end]``, or None."""
    present = [v for v in values[start : end + 1] if v is not None]
    return min(present) if present else None


def _window_max(values: list[float | None], start: int, end: int) -> float | None:
    """Largest present value in ``[start, end]``, or None."""
    present = [v for v in values[start : end + 1] if v is not None]
    return max(present) if present else None


def _window_max_gap(
    left: list[float | None],
    right: list[float | None],
    start: int,
    end: int,
) -> float | None:
    """Largest ``|left - right|`` over frames in ``[start, end]``, or None.

    Only frames where both sides are present contribute; the difference is taken
    within each frame, so it reflects a same-instant left/right comparison rather
    than the gap between two values measured at different times.
    """
    gaps = [
        abs(a - b)
        for a, b in zip(left[start : end + 1], right[start : end + 1], strict=True)
        if a is not None and b is not None
    ]
    return max(gaps) if gaps else None

File: frontend/src/lib/analysis/reps.ts
/**
 * Repetition detection — the TS mirror of `reps.py`.
 *
 * A hysteresis state machine over the hip-height signal. Two thresholds, a lower
 * one to enter the descent and a higher one to close the rep, so the signal must
 * cross a band to advance and a wobble at the boundary cannot register many reps.
 *
 * The batch `detectReps` derives its thresholds from the 90th/10th percentiles
 * of the whole clip, exactly like Python. The state machine itself is factored
 * into `RepStateMachine`, which is already incremental — the live driver feeds
 * it one frame at a time with thresholds set from calibration instead.
 */

import type { AnalysisConfig } from "./config";
import { linearScale } from "./geometry";
import { percentile } from "./smoothing";
import {
  type AngleSeries,
  type Rep,
  type Signal,
} from "./types";

type State = "standing" | "descending" | "ascending";

export interface RepTriple {
  start: number;
  bottom: number;
  end: number;
}

/**
 * The incremental core of rep detection. `push` one frame value at a time; it
 * returns a completed `(start, bottom, end)` triple the instant a rep closes,
 * or null. Untracked frames (null) do not advance it, so a brief dropout does
 * not split a rep in two.
 */
export class RepStateMachine {
  private state: State = "standing";
  private startFrame = 0;
  private bottomFrame = 0;
  private bottomValue = Infinity;
  private lastStandingFrame = 0;

  /**
   * @param turnaroundBand How far the hip must reverse, in torso lengths,
   *   before the bottom counts as passed. Hysteresis, for the same reason the
   *   descend and ascend thresholds are separated: the signal is nearly flat at
   *   the bottom of a rep, so reacting to any flicker makes the phase oscillate
   *   on real tracking noise. Defaults to zero so existing callers and the
   *   parity tests keep their exact previous behaviour.
   */
  constructor(
    private descendThreshold: number,
    private ascendThreshold: number,
    private readonly turnaroundBand = 0,
  ) {}

  /** Update the thresholds mid-stream (the live driver adapts them per rep). */
  setThresholds(descend: number, ascend: number): void {
    this.descendThreshold = descend;
    this.ascendThreshold = ascend;
  }

  /** The current phase, for a live state readout. */
  get phase(): State {
    return this.state;
  }

  push(index: number, value: number | null): RepTriple | null {
    if (value === null) return null;

    if (this.state === "standing") {
      if (value >= this.ascendThreshold) {
        this.lastStandingFrame = index;
      } else if (value < this.descendThreshold) {
        this.state = "descending";
        this.startFrame = this.lastStandingFrame;
        this.bottomFrame = index;
        this.bottomValue = value;
      }
      return null;
    }

    if (this.state === "descending") {
      if (value < this.bottomValue) {
        this.bottomValue = value;
        this.bottomFrame = index;
      } else if (value > this.bottomValue + this.turnaroundBand) {
        // The hip has started back up: the bottom is behind us.
        this.state = "ascending";
      }
      return null;
    }

    // ascending
    if (value < this.bottomValue - this.turnaroundBand) {
      // Dipped again before locking out, so still the same descent. The band
      // keeps a noisy sample from reopening a descent that is genuinely over.
      this.bottomValue = value;
      this.bottomFrame = index;
      this.state = "descending";
      return null;
    }
    if (value >= this.ascendThreshold) {
      const triple: RepTriple = {
        start: this.startFrame,
        bottom: this.bottomFrame,
        end: index,
      };
      this.state = "standing";
      this.lastStandingFrame = index;
      this.bottomValue = Infinity;
      return triple;
    }
    return null;
  }
}

/** Find every repetition in a clip. Empty when there is no detectable movement. */
export function detectReps(angles: AngleSeries, config: AnalysisConfig): Rep[] {
  const signal = angles.hipHeight;
  if (signal.length === 0) return [];

  const baseline = percentile(signal, 90);
  const bottomReference = percentile(signal, 10);
  if (baseline === null || bottomReference === null) return [];

  const travel = baseline - bottomReference;
  if (travel < config.min_rep_range) return [];

  const descend = baseline - config.rep_descent_fraction * travel;
  const ascend = baseline - config.rep_ascent_fraction * travel;

  const machine = new RepStateMachine(descend, ascend, config.rep_turnaround_band);
  const triples: RepTriple[] = [];
  signal.forEach((value, index) => {
    const triple = machine.push(index, value);
    if (triple) triples.push(triple);
  });

  return finalise(triples, angles, config);
}

/** Turn frame triples into `Rep`s, discarding implausibly fast candidates. */
export function finalise(
  triples: RepTriple[],
  angles: AngleSeries,
  config: AnalysisConfig,
): Rep[] {
  const reps: Rep[] = [];
  for (const { start, bottom, end } of triples) {
    const startTime = angles.timestampsS[start];
    const endTime = angles.timestampsS[end];
    if (endTime - startTime < config.min_rep_duration_s) continue;
    reps.push(buildRep(reps.length + 1, start, bottom, end, angles, config));
  }
  return reps;
}

/** Measure one repetition over its frame window. */
export function buildRep(
  index: number,
  start: number,
  bottom: number,
  end: number,
  angles: AngleSeries,
  config: AnalysisConfig,
): Rep {
  const leftMin = windowMin(angles.leftKneeDeg, start, end);
  const rightMin = windowMin(angles.rightKneeDeg, start, end);
  const kneeValues = [leftMin, rightMin].filter((v): v is number => v !== null);
  const minKnee = kneeValues.length ? Math.min(...kneeValues) : null;

  // Asymmetry needs a front-on camera and is the worst instantaneous gap across
  // the rep, not the gap between each leg's independent minimum (that cancels a
  // real one-sided shift). Matches the fix made on the backend.
  const asymmetry =
    angles.view === "front"
      ? windowMaxGap(angles.leftKneeDeg, angles.rightKneeDeg, start, end)
      : null;

  const depthPercent =
    minKnee !== null
      ? linearScale(
          minKnee,
          config.standing_knee_angle_deg,
          config.parallel_knee_angle_deg,
        ) * 100
      : null;

  const offset = windowMax(angles.hipKneeOffset, start, end);

  return {
    index,
    startFrame: start,
    bottomFrame: bottom,
    endFrame: end,
    startTimeS: angles.timestampsS[start],
    bottomTimeS: angles.timestampsS[bottom],
    endTimeS: angles.timestampsS[end],
    minKneeAngleDeg: minKnee,
    minLeftKneeDeg: leftMin,
    minRightKneeDeg: rightMin,
    minHipAngleDeg: windowMin(angles.hipDeg, start, end),
    maxTorsoLeanDeg: windowMax(angles.torsoLeanDeg, start, end),
    kneeAsymmetryDeg: asymmetry,
    depthPercent,
    hipBelowKnee: offset !== null && offset >= 0,
  };
}

export function windowMin(values: Signal, start: number, end: number): number | null {
  let min: number | null = null;
  for (let i = start; i <= end && i < values.length; i++) {
    const value = values[i];
    if (value !== null && (min === null || value < min)) min = value;
  }
  return min;
}

export function windowMax(values: Signal, start: number, end: number): number | null {
  let max: number | null = null;
  for (let i = start; i <= end && i < values.length; i++) {
    const value = values[i];
    if (value !== null && (max === null || value > max)) max = value;
  }
  return max;
}

/** Largest |a - b| over frames where both are present, or null. */
export function windowMaxGap(
  a: Signal,
  b: Signal,
  start: number,
  end: number,
): number | null {
  let max: number | null = null;
  for (let i = start; i <= end && i < a.length && i < b.length; i++) {
    const av = a[i];
    const bv = b[i];
    if (av !== null && bv !== null) {
      const gap = Math.abs(av - bv);
      if (max === null || gap > max) max = gap;
    }
  }
  return max;
}
